### ISSUE TITLE
Invalid SQL Generated for CTE with Count and Distinct

### ISSUE THREAD

**Issue by hahn-kev** [linq2db/linq2db#1380]
When running the following test (setup in `CteTests`) against sql server and SQLite Invalid SQL is generated
```C#
[Test]
public void Test5([CteContextSource] string context)
{
	using (var db = GetDataContext(context))
	{
		var cte1 = db.GetTable<Child>()
			.Where(c => c.ParentID > 1)
			.Select(child => new
			{
				child.ParentID,
				child.ChildID
			}).Distinct()
			.AsCte();
		var query = from p in db.Parent
			join c in cte1 on p.ParentID equals c.ParentID
			join c2 in cte1 on p.ParentID equals c2.ParentID
			select p;

		var cte1_ = db.GetTable<Child>().Where(c => c.ParentID > 1).Select(child => new
		{
			child.ParentID,
			child.ChildID
		}).Distinct();

		var expected =
			from p in db.Parent
			join c in cte1_ on p.ParentID equals c.ParentID
			join c2 in cte1_ on p.ParentID equals c2.ParentID
			select p;

		Assert.AreEqual(expected.Count(), query.Count());
	}
}
```

the SQL generated for the CTE query is as follows
```SQL
-- SqlServer.2012
WITH [CTE_1] ([ParentID])
AS
(
	SELECT DISTINCT
		[t1].[ParentID],
		[t1].[ChildID]
	FROM
		[Child] [t1]
	WHERE
		[t1].[ParentID] > 1
)
SELECT
	Count(*) as [cnt]
FROM
	[Parent] [p]
		INNER JOIN [CTE_1] [c1] ON [p].[ParentID] = [c1].[ParentID]
		INNER JOIN [CTE_1] [c2] ON [p].[ParentID] = [c2].[ParentID]
```
As you can see, CTE_1 only declares `ParentID` as CTE columns, but `ChildID` is also selected because it has the `DISTINCT` clause, it looks like the CTE isn't respecting the changes caused by the `DISTINCT` clause. 

As a work around I can not use a CTE for now.

```
Exception message:
System.Data.SqlClient.SqlException : 'CTE_1' has more columns than were specified in the column list.
```

### Steps to reproduce
Copy paste the snippet above into `CteTests` and run.

### Environment details
linq2db version: *2.5.0*
Database Server: *SqlServer*

**Comment by hahn-kev** [comment]
@sdanyliv I checked out that branch and ran my test against that, same result. I can confirm that the new tests from that branch were ran and did pass. 

I think my issue specifically has to do with calling `Distinct` if I don't do that then it passes just fine.

**Comment by sdanyliv** [comment]
Will check. I'll fix this today.
@MaceWindu, I think we can prepare bug-fix release after merging this PR.

**Comment by sdanyliv** [comment]
@hahn-kev, i have done with issue. Wait for next release or use code from the PR.

**Comment by MaceWindu** [comment]
Should be fixed by #1376
